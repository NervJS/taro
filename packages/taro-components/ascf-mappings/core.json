{
  "Button": {
    "size": {
      "ascfProp": "dimension",
      "valueMap": {
        "mini": "XS",
        "small": "S",
        "default": "M",
        "large": "L"
      },
      "required": false,
      "default": "M",
      "docLink": "https://ascf.docs/components/button#dimension"
    },
    "loading": {
      "ascfProp": "isLoading",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/button#isloading"
    }
  },
  "Input": {
    "password": {
      "ascfProp": "secretInput",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/input#secretinput"
    }
  }
}
