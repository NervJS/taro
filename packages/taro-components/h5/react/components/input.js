import React from 'react'
import reactifyWc from '../utils/reactify-wc'
const Input = reactifyWc('taro-input-core')

// eslint-disable-next-line
const h = React.createElement

export default React.forwardRef((props, ref) => {
  const args = { ...props }

  if (args.hasOwnProperty('focus')) {
    args.autoFocus = Boolean(args.focus)
    delete args.focus
  }

  return (
    React.createElement(Input, { ...args, ref: ref })
  )
})
