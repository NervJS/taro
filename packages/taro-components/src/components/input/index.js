import Nerv from 'nervjs'
import omit from 'omit.js'
import classNames from 'classnames'

import './index.scss'

function getTrueType (type, confirmType, password) {
  if (!type) {
    throw new Error('unexpected type')
  }
  if (confirmType === 'search') type = 'search'
  if (password) type = 'password'
  if (type === 'digit') type = 'number'

  return type
}

function fixControlledValue (value) {
  if (typeof value === 'undefined' || value === null) {
    return ''
  }
  return value
}

class Input extends Nerv.Component {
  constructor () {
    super(...arguments)
    this.onInput = this.onInput.bind(this)
    this.onFocus = this.onFocus.bind(this)
    this.onBlur = this.onBlur.bind(this)
    this.onKeyDown = this.onKeyDown.bind(this)
    this.handleComposition = this.handleComposition.bind(this)

    // input hook
    this.isOnComposition = false
  }

  componentDidMount () {
    // 修复无法选择文件
    if (this.props.type === 'file') {
      this.inputRef.addEventListener('change', this.onInput)
    }
  }

  componentWillUnMount () {
    // 修复无法选择文件
    if (this.props.type === 'file') {
      this.inputRef.removeEventListener('change', this.onInput)
    }
  }

  onInput (e) {
    const {
      type,
      maxLength,
      confirmType,
      password,
      onInput = '',
      onChange = ''
    } = this.props
    if (!this.isOnComposition) {
      let value = e.target.value
      const inputType = getTrueType(type, confirmType, password)
      /* 修复 number 类型 maxLength 无效 */
      if (inputType === 'number' && value && maxLength <= value.length) {
        value = value.substring(0, maxLength)
        e.target.value = value
      }

      Object.defineProperty(e, 'detail', {
        enumerable: true,
        value: { value }
      })
      // 修复 IOS 光标跳转问题
      if (!['number', 'file'].includes(inputType)) {
        const pos = e.target.selectionEnd
        setTimeout(
          () => {
            e.target.selectionStart = pos
            e.target.selectionEnd = pos
          }
        )
      }

      if (onChange) return onChange(e)
      if (onInput) return onInput(e)
    }
  }

  onFocus (e) {
    const { onFocus } = this.props
    Object.defineProperty(e, 'detail', {
      enumerable: true,
      value: {
        value: e.target.value
      }
    })
    onFocus && onFocus(e)
  }

  onBlur (e) {
    const { onBlur } = this.props
    Object.defineProperty(e, 'detail', {
      enumerable: true,
      value: {
        value: e.target.value
      }
    })
    onBlur && onBlur(e)
  }

  onKeyDown (e) {
    const { onConfirm } = this.props
    if (e.keyCode === 13 && onConfirm) {
      Object.defineProperty(e, 'detail', {
        enumerable: true,
        value: {
          value: e.target.value
        }
      })
      onConfirm(e)
    }
  }

  handleComposition (e) {
    if (!(e.target instanceof HTMLInputElement)) return

    if (e.type === 'compositionend') {
      this.isOnComposition = false
    } else {
      this.isOnComposition = true
    }
    this.onInput(e)
  }

  render () {
    const {
      className = '',
      placeholder,
      type = 'text',
      password,
      disabled,
      maxLength,
      confirmType = '',
      focus = false,
      value
    } = this.props
    const cls = classNames('weui-input', className)

    const otherProps = omit(this.props, [
      'className',
      'placeholder',
      'disabled',
      'max',
      'onChange',
      'onFocus',
      'onBlur',
      'type',
      'focus'
    ])

    if ('value' in this.props) {
      otherProps.value = fixControlledValue(value)
    }

    return (
      <input
        ref={input => {
          this.inputRef = input
          input && focus && input.focus()
        }}
        {...otherProps}
        className={cls}
        placeholder={placeholder}
        disabled={disabled}
        max={maxLength}
        onInput={this.onInput}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        autofocus={focus}
        onKeyDown={this.onKeyDown}
        type={getTrueType(type, confirmType, password)}
        onCompositionStart={this.handleComposition}
        onCompositionEnd={this.handleComposition}
      />
    )
  }
}

Input.defaultProps = {
  type: 'text'
}

export default Input
