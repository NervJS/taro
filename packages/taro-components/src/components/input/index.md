input

## API

|              | 属性              | 类型        | 默认值 | 说明                                                                                                            |
| ------------ | ----------------- | ----------- | ------ | --------------------------------------------------------------------------------------------------------------- |
| √            | value             | String      |        | 输入框的初始内容                                                                                                |
| √            | type              | String      | 23     | input 的类型                                                                                                    |
| √            | password          | Boolean       |        | 是否是密码类型                                                                                                  |
| √            | placeholder       | String      |        | 输入框为空时占位符                                                                                              |
|              | placeholder-style | String      |        | 指定 placeholder 的样式                                                                                         |
|              | placeholder-class | String      |        | 指定 placeholder 的样式类                                                                                       |
| √            | disabled          | Boolean     |        | 是否禁用                                                                                                        |
| √            | maxLength         | Number      |        | 最大输入长度，设置为 -1 的时候不限制最大长度                                                                    |
|              | cursor-spacing    | Number      |        | 指定光标与键盘的距离，单位 px 。取 input 距离底部的距离和 cursor-spacing 指定的距离的最小值作为光标与键盘的距离 |
|              | auto-focus        | Boolean     |        | (即将废弃，请直接使用 focus )自动聚焦，拉起键盘                                                                 |
|              | focus             | Boolean     |        | 获取焦点                                                                                                        |
|     √           | confirm-type      | String      |        | 设置键盘右下角按钮的文字(目前仅支持search)                                                                                        |
|              | cursor            | Number      |        | 指定 focus 时的光标位置                                                                                         |
|              | selection-start   | Number      |        | 光标起始位置，自动聚集时有效，需与 selection-end 搭配使用                                                       |
|              | selection-end     | Number      |        | 光标结束位置，自动聚集时有效，需与 selection-start 搭配使用                                                     |
|              | adjust-position   | Boolean     |        | 键盘弹起时，是否自动上推页面                                                                                    |
| √ (onChange) | bindinput         | EventHandle |        | 当键盘输入时，触发 input 事件，处理函数可以直接 return 一个字符串，将替换输入框的内容。                         |
| √ (onFocus)  | bindfocus         | EventHandle |        | 输入框聚焦时触发，height 参数在基础库 1.9.90 起支持                                                             |
| √ (onBlur)   | bindblur          | EventHandle |        | 输入框失去焦点时触发                                                                                            |
|              | bindconfirm       | EventHandle |        | 点击完成按钮时触发                                                                                              |
