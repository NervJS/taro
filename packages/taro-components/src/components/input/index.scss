input {
  display: block;
  height: 1.4rem;
  text-align: inherit;
  text-overflow: clip;
  overflow: hidden;
  white-space: nowrap;
}
