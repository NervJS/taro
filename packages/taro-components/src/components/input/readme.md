# taro-input



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description | Type      | Default     |
| ------------- | -------------- | ----------- | --------- | ----------- |
| `confirmType` | `confirm-type` |             | `string`  | `undefined` |
| `disabled`    | `disabled`     |             | `boolean` | `false`     |
| `maxlength`   | `max-length`   |             | `number`  | `undefined` |
| `password`    | `password`     |             | `string`  | `undefined` |
| `placeholder` | `placeholder`  |             | `string`  | `undefined` |
| `type`        | `type`         |             | `string`  | `'text'`    |
| `value`       | `value`        |             | `string`  | `undefined` |


## Events

| Event     | Description | Type               |
| --------- | ----------- | ------------------ |
| `blur`    |             | `CustomEvent<any>` |
| `change`  |             | `CustomEvent<any>` |
| `confirm` |             | `CustomEvent<any>` |
| `focus`   |             | `CustomEvent<any>` |
| `input`   |             | `CustomEvent<any>` |
| `keydown` |             | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
