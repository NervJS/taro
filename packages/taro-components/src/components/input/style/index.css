input {
  display: block;
  height: 24px;
  text-align: inherit;
  text-overflow: clip;
  overflow: hidden;
  white-space: nowrap;
}
