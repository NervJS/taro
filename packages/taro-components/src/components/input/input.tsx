import { Component, h, ComponentInterface, Prop, Event, EventEmitter, Element, State, Watch } from '@stencil/core'
import { EventHandler, TaroEvent } from '../../../types'

function getTrueType (type: string | undefined, confirmType: string, password: boolean) {
  if (confirmType === 'search') type = 'search'
  if (password) type = 'password'
  if (typeof type === 'undefined') {
    return 'text';
  }
  if (!type) {
    throw new Error('unexpected type')
  }
  if (type === 'digit') type = 'number'

  return type
}

function fixControlledValue (value?: string) {
  return value ?? ''
}

@Component({
  tag: 'taro-input-core',
  styleUrl: 'index.scss'
})
export class Input implements ComponentInterface {
  private inputRef: HTMLInputElement
  private isOnComposition = false
  private onInputExcuted = false
  private fileListener: EventHandler

  @Prop() value: string
  @Prop() type: string
  @Prop() password = false
  @Prop() placeholder: string
  @Prop() disabled = false
  @Prop() maxlength = 140
  @Prop() autoFocus = false
  @Prop() confirmType = 'done'
  @Prop() name: string

  @State() _value: string

  @Element() el: HTMLElement

  @Watch('value')
  watchHandler (newValue: string, oldValue: string) {
    if (newValue !== oldValue) {
      this._value = newValue
    }
  }

  @Watch('autoFocus')
  watchFocus (newValue: boolean, oldValue: boolean) {
    if (!oldValue && newValue) {
      this.inputRef.focus()
    }
  }

  @Event({
    eventName: 'input'
  }) onInput: EventEmitter

  @Event({
    eventName: 'focus'
  }) onFocus: EventEmitter

  @Event({
    eventName: 'blur'
  }) onBlur: EventEmitter

  @Event({
    eventName: 'confirm'
  }) onConfirm: EventEmitter

  @Event({
    eventName: 'change'
  }) onChange: EventEmitter

  @Event({
    eventName: 'keydown'
  }) onKeyDown: EventEmitter

  componentWillLoad () {
    this._value = this.value
  }

  componentDidLoad () {
    if (this.type === 'file') {
      this.fileListener = () => {
        this.onInput.emit()
      }
      this.inputRef.addEventListener('change', this.fileListener)
    } else {
      this.inputRef.addEventListener('compositionstart', this.handleComposition)
      this.inputRef.addEventListener('compositionend', this.handleComposition)
    }

    Object.defineProperty(this.el, 'value', {
      get: () => this.inputRef.value,
      set: value => {
        this._value = value
      },
      configurable: true
    })
  }

  disconnectedCallback () {
    if (this.type === 'file') {
      this.inputRef.removeEventListener('change', this.fileListener)
    }
  }

  handleInput = (e: TaroEvent<HTMLInputElement>) => {
    e.stopPropagation()
    const {
      type,
      maxlength,
      confirmType,
      password
    } = this
    if (!this.isOnComposition && !this.onInputExcuted) {
      let value = e.target.value
      const inputType = getTrueType(type, confirmType, password)
      this.onInputExcuted = true
      /* 修复 number 类型 maxlength 无效 */
      if (inputType === 'number' && value && maxlength <= value.length) {
        value = value.substring(0, maxlength)
        e.target.value = value
      }

      // 修复 IOS 光标跳转问题
      // if (!(['number', 'file'].indexOf(inputType) >= 0)) {
      //   const pos = e.target.selectionEnd
      //   setTimeout(
      //     () => {
      //       e.target.selectionStart = pos
      //       e.target.selectionEnd = pos
      //     }
      //   )
      // }

      this._value = value

      this.onInput.emit({
        value,
        cursor: value.length
      })
    }
  }

  handleFocus = (e: TaroEvent<HTMLInputElement> & FocusEvent) => {
    this.onInputExcuted = false
    this.onFocus.emit({
      value: e.target.value
    })
  }

  handleBlur = (e: TaroEvent<HTMLInputElement> & FocusEvent) => {
    this.onBlur.emit({
      value: e.target.value
    })
  }

  handleChange = (e: TaroEvent<HTMLInputElement>) => {
    e.stopPropagation()
    this.onChange.emit({
      value: e.target.value
    })
  }

  handleKeyDown = (e: TaroEvent<HTMLInputElement> & KeyboardEvent) => {
    const { value } = e.target
    this.onInputExcuted = false
    e.stopPropagation()

    this.onKeyDown.emit({ value })

    e.keyCode === 13 && this.onConfirm.emit({ value })
  }

  handleComposition = (e) => {
    if (!(e.target instanceof HTMLInputElement)) return

    if (e.type === 'compositionend') {
      this.isOnComposition = false
      this.onInput.emit({ value: e.target.value })
    } else {
      this.isOnComposition = true
    }
  }

  render () {
    const {
      _value,
      type,
      password,
      placeholder,
      disabled,
      maxlength,
      autoFocus,
      confirmType,
      name
    } = this

    return (
      <input
        ref={input => {
          this.inputRef = input!
          autoFocus && input?.focus()
        }}
        class='weui-input'
        value={fixControlledValue(_value)}
        type={getTrueType(type, confirmType, password)}
        placeholder={placeholder}
        disabled={disabled}
        maxlength={maxlength}
        autofocus={autoFocus}
        name={name}
        onInput={this.handleInput}
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}
        onChange={this.handleChange}
        onKeyDown={this.handleKeyDown}
      />
    )
  }
}
