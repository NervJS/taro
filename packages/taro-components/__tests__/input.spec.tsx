import { h } from '@stencil/core'
import { newSpecPage, SpecPage } from '@stencil/core/testing'

import { Input } from '../src/components/input/input'

describe('Input', () => {
  let page: SpecPage

  it('props', async () => {
    let maxLength = 10
    const disabled = false
    let value = ''
    const placeholder = 'placeholder'
    const autoFocus = true
    let type = 'text'
    const onFocus = jest.fn()

    page = await newSpecPage({
      components: [Input],
      template: () => (<taro-input-core
        maxlength={maxLength}
        disabled={disabled}
        value={value}
        placeholder={placeholder}
        type={type}
        autoFocus={autoFocus}
        onFocus={onFocus}
      />),
    })
    const input = page.root?.querySelector('input')
    expect(input?.classList.contains('weui-input')).toEqual(true)
    expect(input?.getAttribute('maxlength')).toEqual(maxLength.toString())
    expect(input?.getAttribute('placeholder')).toEqual(placeholder)
    expect(input?.type).toEqual(type)

    input?.focus()
    expect(onFocus).toHaveBeenCalledTimes(1)
    // assert($(input).is(':focus'))
    // assert(document.activeElement === input)

    value = 'test'
    type = 'password'
    maxLength = 4
    page.root?.setAttribute('value', value)
    page.root?.setAttribute('maxlength', maxLength.toString())
    page.root?.setAttribute('type', type)
    // @ts-ignore
    input.value = value
    input?.dispatchEvent(new Event('input'))
    await page.waitForChanges()

    expect(page.root?.getAttribute('value')).toEqual(value)
    expect(input?.type).toEqual(type)

    expect(page.root).toMatchSnapshot()
  })

  it('event', async () => {
    const maxLength = 4
    const value = 'value'
    const onFocus = jest.fn()
    const onBlur = jest.fn()
    const onKeyDown = jest.fn()
    const onConfirm = jest.fn()
    const onInput = jest.fn()
    page = await newSpecPage({
      components: [Input],
      template: () => (<taro-input-core
        maxlength={maxLength}
        value={value}
        onFocus={onFocus}
        onBlur={onBlur}
        onKeyDown={onKeyDown}
        onConfirm={onConfirm}
        onInput={onInput}
      />),
    })
    const input = page.root?.querySelector('input')

    input?.focus()
    expect(onFocus).toHaveBeenCalledTimes(1)
    expect(onFocus.mock.calls[0][0].detail).toEqual({ value })

    input?.blur()
    expect(onBlur).toHaveBeenCalledTimes(1)
    expect(onBlur.mock.calls[0][0].detail).toEqual({ value })

    input?.dispatchEvent(new KeyboardEvent('keydown', { key: 'Delete', keyCode: 46 }))
    expect(onKeyDown).toHaveBeenCalledTimes(1)
    expect(onKeyDown.mock.calls[0][0].detail).toEqual({ cursor: value.length, keyCode: 46, value })
    expect(onConfirm).toHaveBeenCalledTimes(0)

    input?.dispatchEvent(new KeyboardEvent('keydown', { key: 'e', keyCode: 69 }))
    expect(onKeyDown).toHaveBeenCalledTimes(2)
    expect(onKeyDown.mock.calls[1][0].detail).toEqual({ cursor: value.length, keyCode: 69, value })
    expect(onConfirm).toHaveBeenCalledTimes(0)

    input?.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 13 }))
    expect(onConfirm).toHaveBeenCalledTimes(1)
    const inputEvent = {
      data: 'a',
      inputType: 'insertText'
    }
    // @ts-ignore
    input?.dispatchEvent(new CustomEvent('input', inputEvent))
    expect(onInput).toHaveBeenCalledTimes(1)

    expect(page.root).toMatchSnapshot()
  })
})
