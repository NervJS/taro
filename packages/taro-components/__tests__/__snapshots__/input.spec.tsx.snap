// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input event 1`] = `
<taro-input-core>
  <input class="weui-input" maxlength="4" type="text" value="value">
</taro-input-core>
`;

exports[`Input props 1`] = `
<taro-input-core maxlength="4" type="password" value="test">
  <input autofocus="" class="weui-input" maxlength="4" placeholder="placeholder" type="password" value="test">
</taro-input-core>
`;
